syncapi/types/types.go

```diff
@@ -27,138 +27,213 @@ import (
+	return t.Positions[1]
+}
+
+// IsAfter returns true if ANY position in this token is greater than `other`.
+func (t *StreamingToken) IsAfter(other StreamingToken) bool {
+	for i := range other.Positions {
+		if t.Positions[i] > other.Positions[i] {
+			return true
+		}
+	}
+	return false
+}
+
+// WithUpdates returns a copy of the StreamingToken with updates applied from another StreamingToken.
+// If the latter StreamingToken contains a field that is not 0, it is considered an update,
+// and its value will replace the corresponding value in the StreamingToken on which WithUpdates is called.
+func (t *StreamingToken) WithUpdates(other StreamingToken) (ret StreamingToken) {
+	ret.Type = t.Type
+	ret.Positions = make([]StreamPosition, len(t.Positions))
+	for i := range t.Positions {
+		ret.Positions[i] = t.Positions[i]
+		if other.Positions[i] == 0 {
+			continue
+		}
+		ret.Positions[i] = other.Positions[i]
+	}
+	return ret
+}
+
+type TopologyToken struct {
+	syncToken
+}
+
+func (t *TopologyToken) Depth() StreamPosition {
+	return t.Positions[0]
+}
+func (t *TopologyToken) PDUPosition() StreamPosition {
+	return t.Positions[1]
+}
+func (t *TopologyToken) StreamToken() StreamingToken {
+	return NewStreamToken(t.PDUPosition(), 0)
+}
+func (t *TopologyToken) String() string {
+	return t.syncToken.String()
+}
+
+// Decrement the topology token to one event earlier.
+func (t *TopologyToken) Decrement() {
+	depth := t.Positions[0]
+	pduPos := t.Positions[1]
+	if depth-1 <= 0 {
+		depth = 1
+	} else {
+		depth--
+		pduPos += 1000
+	}
+	// The lowest token value is 1, therefore we need to manually set it to that
+	// value if we're below it.
+	if depth < 1 {
+		depth = 1
+	}
+	t.Positions = []StreamPosition{
+		depth, pduPos,
+	}
 }
 
-// NewPaginationTokenFromString takes a string of the form "xyyyy..." where "x"
+// NewSyncTokenFromString takes a string of the form "xyyyy..." where "x"
 // represents the type of a pagination token and "yyyy..." the token itself, and
-// parses it in order to create a new instance of PaginationToken. Returns an
+// parses it in order to create a new instance of SyncToken. Returns an
 // error if the token couldn't be parsed into an int64, or if the token type
-// isn't a known type (returns ErrInvalidPaginationTokenType in the latter
+// isn't a known type (returns ErrInvalidSyncTokenType in the latter
 // case).
-func NewPaginationTokenFromString(s string) (token *PaginationToken, err error) {
+func newSyncTokenFromString(s string) (token *syncToken, err error) {
 	if len(s) == 0 {
-		return nil, ErrInvalidPaginationTokenLen
+		return nil, ErrInvalidSyncTokenLen
 	}
 
-	token = new(PaginationToken)
+	token = new(syncToken)
 	var positions []string
 
-	switch t := PaginationTokenType(s[:1]); t {
-	case PaginationTokenTypeStream, PaginationTokenTypeTopology:
+	switch t := SyncTokenType(s[:1]); t {
+	case SyncTokenTypeStream, SyncTokenTypeTopology:
 		token.Type = t
 		positions = strings.Split(s[1:], "_")
```

Comment:
It doesn't appear as if we guard anywhere against the sync token containing no positions, which becomes a problem later on in `Depth()`, `PDUPosition()` and `EDUPosition()` etc where we index on specific array indices. That could result in panics, so it's probably best to make sure we check that we get at least the number of expected positions *somewhere*.